src/main/java/de/tum/in/www1/artemis/config/migration/setups/localvc/gitlab/MigrationEntryGitLabToLocalVC.java

```diff
@@ -0,0 +1,321 @@
+                errorList.add(solutionParticipation);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void migrateTemplates(List<TemplateProgrammingExerciseParticipation> templateParticipations) {
+        for (var templateParticipation : templateParticipations) {
+            try {
+                if (isRepositoryUriNotNull(templateParticipation, "Repository URI is null for template participation with id {}, cant migrate")) {
+                    var url = cloneRepositoryFromSourceVCSAndMoveToLocalVC(templateParticipation.getProgrammingExercise(), templateParticipation.getRepositoryUri(),
+                            templateParticipation.getProgrammingExercise().getBranch());
+                    if (url == null) {
+                        log.error("Failed to migrate template repository for template participation with id {}, keeping the url in the database", templateParticipation.getId());
+                        errorList.add(templateParticipation);
+                    }
+                    else {
+                        log.debug("Migrated template participation with id {} to {}", templateParticipation.getId(), url);
+                        templateParticipation.setRepositoryUri(url);
+                        templateProgrammingExerciseParticipationRepository.save(templateParticipation);
+                    }
+                }
+            }
+            catch (Exception e) {
+                log.error("Failed to migrate template participation with id {}", templateParticipation.getId(), e);
+                errorList.add(templateParticipation);
+            }
+        }
+    }
+
+    /**
+     * Migrate the student participations. This is the most time-consuming part of the migration as we have
+     * to clone the repository for each student.
+     *
+     * @param participations list of student participations to migrate
+     */
+    @Override
+    protected void migrateStudents(List<ProgrammingExerciseStudentParticipation> participations) {
+        for (var participation : participations) {
+            try {
+                if (isRepositoryUriNotNull(participation, "Repository URI is null for student participation with id {}, cant migrate")) {
+                    var url = cloneRepositoryFromSourceVCSAndMoveToLocalVC(participation.getProgrammingExercise(), participation.getRepositoryUri(), participation.getBranch());
+                    if (url == null) {
+                        log.error("Failed to migrate student repository for student participation with id {}, keeping the url in the database", participation.getId());
+                        errorList.add(participation);
+                    }
+                    else {
+                        log.debug("Migrated student participation with id {} to {}", participation.getId(), url);
+                        if (participation.getBranch() != null) {
```

Comment:
Should this be a `!defaultBranch.equals(participation.getBranch()` instead?

Reply:
I don't know what `null` for a branch means here. This [part](https://github.com/ls1intum/Artemis/blob/chore/bitbucket-localvc-migration-release/src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20240103_143700.java) is from the [BitBucket to LocalVC migration](https://github.com/ls1intum/Artemis/pull/8122). @reschandreas What meaning has the `null` here? Is that intended or not?

Reply:
I don't recall; it's been a while since I've worked with this codebase.

Reply:
Okay, then would assume that it should be compared with the `defaultBranch`, similar to the other situations in the code with `setBranch`. Fixed with 6945886276468abb9cbced8ef753bb5aaa51a9bb